rfc/rfc-0000-supply-chain-tracing.md

```diff
@@ -0,0 +1,181 @@
+    outputs:
+    - name:                  # string
+      digest:                # string
+      path:                  # string
+      lastTransitionTime:    # metav1.Time
+```
+
+Based on the [basic supply chain example](https://github.com/vmware-tanzu/cartographer/tree/53402edb8b8914b4cd36ace82da85f83f3daefc1/examples/basic-sc), the Workload status would have the form:
+
+```yaml
+... existing fields ...
+status:
+  ... existing fields ...
+  resources:
+  - name: source-provider
+    stampedRef:
+      apiVersion: source.toolkit.fluxcd.io/v1beta1
+      kind: GitRepository
+      namespace: default
+      name: my-workload
+    templateRef:
+      apiVersion: carto.run/v1
+      kind: ClusterSourceTemplate
+      name: source
+    outputs:
+    - name: url
+      digest: sha256:0c20a75353e12f51e4e7f42525d2a0f135b9dacea1648406fd90dac76f46e27c
+      path: .status.artifact.url
+      lastTransitionTime: "2022-02-16T03:29:52Z"
+    - name: revision
+      digest: sha256:76004630fc3135b24df7d411182476d9324e091975f7ef15137fc4c25acd5056
+      path: .status.artifact.revision
+      lastTransitionTime: "2022-02-16T03:29:52Z"
+  - name: image-builder
+    stampedRef:
+      apiVersion: kpack.io/v1alpha2
+      kind: Image
+      namespace: default
+      name: my-workload
+    templateRef:
+      apiVersion: carto.run/v1
+      kind: ClusterImageTemplate
+      name: image
+    inputs:
+    - name: source-provider
+    outputs:
+    - name: image
+      digest: sha256:eeedc5e676261368af5f5ebd6c66d3f853c7deff4cb2966a0621f8f546d77a42
+      path: .status.latestImage
+      lastTransitionTime: "2022-02-16T03:23:37Z"
+  - name: deployer
+    stampedRef:
+      apiVersion: kappctrl.k14s.io/v1alpha1
+      kind: App
+      namespace: default
+      name: my-workload
+    templateRef:
+      apiVersion: carto.run/v1
+      kind: ClusterTemplate
+      name: app-deploy
+    inputs:
+    - name: image-builder
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+- `.status.resources[*]` should match the order and length of `.spec.resource` for a ClusterSupplyChain.
+- `.status.resources[*].name` is the name of the resources as defined within the ClusterSupplyChain resource's array.
+- `.status.resources[*].templateRef` object reference to the template resource referenced by the ClusterSupplyChain.
+- `.status.resources[*].stampedRef` object reference to the Kubernetes resource created by Cartographer for this resource.
+- `.status.resources[*].inputs[*]` inputs model the relationships between resources within the SupplyChain graph. The value of an input can be read from the stamped resource based on the outputs for the referenced resource.
+- `.status.resources[*].inputs[*].name` the name of the resource backing this input. 
+- `.status.resources[*].outputs[*].name` the name of the output. Output names are fixed and defined by the template type.
+- `.status.resources[*].outputs[*].digest` sha256sum of the raw output value.
```

Comment:
I wonder if we need to be more specific here given how "sensitive" hashing is to any bit change, e.g., a javascript based client might end up interpreting the raw value differently from the controller due to a different json decoder?

Reply:
I know you mentioned that we should try as much as possible to not expose potentially sensitive information in `workload.status`, but storage-wise, made me wonder if it'd be too much of a crazy/bad idea considering a custom apiserver implementation to work around the etcd storage limitation

Reply:
>I wonder if we need to be more specific here given how "sensitive" hashing is to any bit change, e.g., a javascript based client might end up interpreting the raw value differently from the controller due to a different json decoder?

That's a good point, the digest will only be stable against a string or byte array value, structs are ambiguous. The intent isn't to be cryptographically secure, but to provide a check of the value so a client can be reasonably sure that it has the "correct" value and that the value hasn't changed between when the c

We can drop the `digest` field for now if you're not confident in it, and add it back once we have the semantics nailed.

> made me wonder if it'd be too much of a crazy/bad idea considering a custom apiserver implementation to work around the etcd storage limitation

yes, crazy and bad :)

But on a serious note, there are other concerns that would be nice to have that will never end up on the Workload status. For example, a history of artifacts and how the interact in the supply chain. This kind of data should be captured into a different system that can support temporal (and other types of) queries. I wouldn't use that front an aggregated api, but to be used as a separate store.

Reply:
I'm not sure how sensible this is, but it is possible to serialize with sorted keys in the go yaml package (unsure about the k8s one). That would at least produce repeatable digests.

Reply:
We can also say that the digest is primarily for Cartographer's internal use (to generate the lastTransitionTime corectly).